docs/HowSwiftImportsCModules.md

```diff
@@ -0,0 +1,922 @@
+be converted to a Swift closure; however, bridging is required to adjust the
+memory layout.
+
+As discussed above, there are cases where bridging that adjusts memory layout is
+not possible, for example, when importing pointers to function pointers. For
+example, while C's int `(*)(char)` can be imported as `(Int8) -> Int` (requires
+an adjustment of memory layout), C's `int (**)(char)` can't be imported as
+`UnsafePointer<(Int8) -> Int>`, because the pointee must have identical memory
+layout in C and in Swift.
+
+Therefore, we need a Swift type that has a memory layout identical to C function
+pointers, at least for such fallback cases. This type is spelled `@convention(c)
+(Arg1, Arg2, Arg3) -> Result`.
+
+Even though it is possible to import C function pointers as Swift closure with a
+context pointer in some cases, C function pointers are always imported as
+`@convention(c)` "closures" (in quotes because they don't have a context
+pointer, so they are not real closures). Swift provides an implicit conversion
+from `@convention(c)` closures to Swift closures with a context.
+
+Importing C function pointers also takes take pointer nullability into account:
+a nullable C function pointer is imported as optional.
+
+```c
+// C header.
+
+void qsort(
+  void *base,
+  size_t nmemb,
+  size_t size,
+  int (*compar)(const void *, const void *));
+
+void qsort_annotated(
+  void * _Nonnull base,
+  size_t nmemb,
+  size_t size,
+  int (* _Nonnull compar)(const void * _Nonnull, const void * _Nonnull));
+```
+
+```swift
+// C header imported in Swift.
+
+func qsort(
+  _ base: UnsafeMutableRawPointer!,
+  _ nmemb: Int32,
+  _ size: Int32,
+  _ compar: (@convention(c) (UnsafeRawPointer?, UnsafeRawPointer?) -> Int32)!
+)
+
+func qsort_annotated(
+  _ base: UnsafeMutableRawPointer,
+  _ nmemb: Int32,
+  _ size: Int32,
+  _ compar: @convention(c) (UnsafeRawPointer, UnsafeRawPointer) -> Int32
+)
+```
+
+See also Apple's documentation about this topic: [Using Imported C Functions in
+Swift](https://developer.apple.com/documentation/swift/imported_c_and_objective-c_apis/using_imported_c_functions_in_swift)
+
+# Fixed-size arrays
+
+C's fixed-size arrays are imported as Swift tuples.
+
+```c
+// C header.
+
+extern int x[4];
+```
+
+```swift
+// C header imported in Swift.
+
+var x: (Int, Int, Int, Int) { get set }
+```
+
+This mapping strategy is widely recognized as being far from optimal because
+ergonomics of Swift tuples does not match C's fixed-size arrays. For example,
+Swift tuples cannot be accessed through an index that only becomes known at
+runtime. One get an unsafe pointer to values in a homogeneous tuple with
+`withUnsafeMutablePointer(&myTuple) { ... }`.
+
+Fixed-size arrays are a commonly requested feature in Swift, and a good proposal
+is likely to be accepted. Once Swift has fixed-size arrays natively in the
+language, we can use them to improve C interoperability.
+
+# Structs
```

Comment:
See also Apple's documentation about this topic: [Using Imported C Structs and Unions in Swift](https://developer.apple.com/documentation/swift/imported_c_and_objective-c_apis/using_imported_c_structs_and_unions_in_swift)

Reply:
Added.